### DISCUSSION TITLE
umi3升级到umi4，ProForm下的<Form.Item>无法获取上下文，报错Can not find FormContext. Please make sure you wrap Field under Form；

### DISCUSSION THREAD

**Discussion by titld** [umijs/umi#10019] [Q&A]
ProForm里面引入自定义组件<Form.Item>，
![image](https://user-images.githubusercontent.com/57819648/206703512-88968ca7-7637-4d5d-9c3c-d08566ead67b.png)
此时没有问题；但是如果我把
![image](https://user-images.githubusercontent.com/57819648/206703771-94c12d8d-df88-4a78-a77e-178c35c9a9d1.png)
这一部分抽出来，封装成一个插件'@umijs/plugin-my-demo'，插件导出MyPlugins组件（真实的插件，从自定的插件中导出），然后以组件的方式引入
```js
import { MyPlugins } from '@umijs/plugin-my-demo';
<ProForm><MyPlugins /></ProForm>
```
就会破坏原有的Form结构，会报错：【Antd】Warning: Can not find FormContext. Please make sure you wrap Field under Form；
在umi3当中没有问题，但在umi4中就会出错而无法使用，是不是umi插件机制有问题，拿不到context？

© 版本信息
ProComponents 版本: 2.3.19
ProForm 版本：2.2.7
umi 版本 4.0.34

**Comment by fz6m** [comment]
给一个最小复现，怎么封装的，具体上下文给一下。

**Reply by titld** [reply]
复现很麻烦，这里说下具体做法：
1.封装一个@chi/data-field组件；
index.ts文件：
```js
import { Form, Input } from "antd";
function dataField() {
return <Form.Item>
<Input />
</Form.Item>
}
export default dataField;
```
package.json文件：
```js
{
  "name": "@chi/data-field",
  "version": "0.1.0",
  "description": "@chi/data-field",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib",
    "src"
  ],
  "peerDependencies": {
    "antd": "4.x",
    "umi": "4.x"
  },
  "publishConfig": {
    "access": "public"
  }
}
```
2.发布包：
将`@chi/data-field`发布到制品库；

3.在新的页面中引入刚刚发布的包；
app.js文件：
```js
import { ProForm } from "@ant-design/pro-form";
import dataField from "@chi/data-field";

function dataField() {
return <ProForm>
<dataField />
</ProForm>
}
```
4.此时就会报错Can not find FormContext. Please make sure you wrap Field under Form；

号外：如果dataField是以一个同项目文件下的组件的形式引入，则不会报错；

**Reply by fz6m** [reply]
需要给最小复现（可以用 monorepo 模拟多包），我尝试了一下，并没有复现你说的问题。

**Reply by titld** [reply]
非常感谢您，找到原因了，本地测试包的时候我使用workspaces链接的packages下的lib文件；今天测试线上包发现是没问题的；
只是还有个疑问，
umi3的时候不会报错，控制台会有一个警告：./node_modules/connected-react-router/esm/ConnectedRouter.js，
"'react-redux' does not contain an export named 'ReactReduxContext'"; 不会影响结果，能正常使用；
升级umi4后，会报错Can not find FormContext. Please make sure you wrap Field under Form；，不能正常使用；

**Reply by fz6m** [reply]
1. 排查下 `FormContext` 是哪里的，这个导入可能在哪个位置使用了，在代码里搜一下。

2. 这个导入是第三方包里使用的吗？本地开发会报错？详细信息再提供一些

3. 升级下 umi 最新版本 `4.0.36` ，关闭 `mfsu: false` ，再检查下有没有问题。
